### ISSUE TITLE
`FromIterator` type inference bug

### ISSUE THREAD

**Issue by fosskers** [coalton-lang/coalton#792]
Hi there, thanks for Coalton. It's a pretty exciting project. In first trying it out, I wanted to test out `Iterator`s, so I threw this together:
```lisp
(coalton-toplevel
  (declare fast-evens (List Integer -> List Integer))
  (define (fast-evens items)
    "Be efficient!"
    (iter:collect! (iter:into-iter items))))
```
but was surprised when it didn't compile. The error given was:
```
error: 
    (during macroexpansion of (COALTON-TOPLEVEL
      (DECLARE FAST-EVENS ...)
      ...))
    Ambigious constraint FROMITERATOR (LIST INTEGER) #T17275
```
A similar thing happens often in Rust when using `collect`, and we're often forced to do something like `.collect::<Vec<_>>()`, but in this Coalton case, I'm using an explicit `declare` so there should be no mystery as to the expected output type of `collect`. Note that _really_ trying to convince the code that the result if a `List` by using a `filter` (etc.) after the `collect!` doesn't work either. 

Any idea what might be happening here? I'm using the latest commit on the `master` branch of this repository. Thank you!

**Comment by colescott** [comment]
This appears to be a bug where functional dependencies are not being resolved for explicit bindings. 

This should be fixed in #793

**Comment by fosskers** [comment]
Thank you!
